libraries/ESP8266WiFi/src/BearSSLHelpers.cpp

```diff
@@ -0,0 +1,830 @@
+      ta->pkey.key.ec.q = (uint8_t*)malloc(pk->key.ec.qlen);
+      if (ta->pkey.key.ec.q == nullptr) {
+        goto cleanup;
+      }
+      memcpy(ta->pkey.key.ec.q, pk->key.ec.q, pk->key.ec.qlen);
+      ta->pkey.key.ec.qlen = pk->key.ec.qlen;
+      return true;
+    default:
+      goto cleanup;
+  }
+  return true;
+
+  // In case of failure, free any allocations and fail out
+cleanup:
+  free_ta_contents(ta);
+  return false;
+}
+
+br_x509_trust_anchor *certificate_to_trust_anchor(const br_x509_certificate *xc)
+{
+  br_x509_trust_anchor *ta = (br_x509_trust_anchor*)malloc(sizeof(br_x509_trust_anchor));
+  if (!ta) return nullptr;
+
+  if (!certificate_to_trust_anchor_inner(ta, xc)) {
+    free(ta);
+    return nullptr;
+  } else {
+    return ta;
+  }
+}
+
+void free_ta_contents(br_x509_trust_anchor *ta)
+{
+  if (ta) {
+    free(ta->dn.data);
+    if (ta->pkey.key_type == BR_KEYTYPE_RSA) {
+      free(ta->pkey.key.rsa.n);
+      free(ta->pkey.key.rsa.e);
+    } else if (ta->pkey.key_type == BR_KEYTYPE_EC) {
+      free(ta->pkey.key.ec.q);
+    }
+    memset(ta, 0, sizeof(*ta));
+  }
+}
+
+// Checks if a bitstream looks like a valid DER(binary) encoding.
+// Basically tries to verify the length of all included segments
+// matches the length of the input buffer.  Does not actually
+// validate any contents.
+bool looks_like_DER(const unsigned char *buff, size_t len)
+{
+  if (len < 2) {
+    return false;
+  }
+  if (pgm_read_byte(buff++) != 0x30) {
+    return false;
+  }
+  int fb = pgm_read_byte(buff++);
+  len -= 2;
+  if (fb < 0x80) {
+    return (size_t)fb == len;
+  } else if (fb == 0x80) {
+    return false;
+  } else {
+    fb -= 0x80;
+    if (len < (size_t)fb + 2) {
+      return false;
+    }
+    len -= (size_t)fb;
+    size_t dlen = 0;
+    while (fb -- > 0) {
+      if (dlen > (len >> 8)) {
+        return false;
+      }
+      dlen = (dlen << 8) + (size_t)pgm_read_byte(buff++);
+    }
+    return dlen == len;
+  }
+}
+
+void free_pem_object_contents(pem_object *po)
+{
+  if (po) {
+    free(po->name);
+    free(po->data);
+  }
+}
+
+// Converts a PEM (~=base64) source into a set of DER-encoded binary blobs.
+// Each blob is named by the ---- BEGIN xxx ---- field, and multiple
+// blobs may be returned.
+pem_object *decode_pem(const void *src, size_t len, size_t *num)
+{
+  std::vector<pem_object> pem_list;
+  std::unique_ptr<br_pem_decoder_context> pc(new br_pem_decoder_context); // auto-delete on exit
+  if (!pc.get()) return nullptr;
+  pem_object po, *pos;
+  const unsigned char *buff;
+  std::vector<uint8_t> bv;
+
+  *num = 0;
+  br_pem_decoder_init(pc.get());
+  buff = (const unsigned char *)src;
+  po.name = nullptr;
+  po.data = nullptr;
+  po.data_len = 0;
+  bool inobj = false;
+  bool extra_nl = true;
+
+  while (len > 0) {
+    size_t tlen;
+
+    tlen = br_pem_decoder_push(pc.get(), buff, len);
+    buff += tlen;
+    len -= tlen;
+    switch (br_pem_decoder_event(pc.get())) {
```

Comment:
no default case for this switch?

Reply:
Will add default and fallthrough to the error case, but by contract we've gone through all enum values it can return.  Anything else means we've had memory corruption somewhere...